pallets/moonbeam-lazy-migrations/src/foreign_asset.rs

```diff
@@ -0,0 +1,221 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! # Functions for handling foreign asset migrations
+
+use super::*;
+use frame_support::sp_runtime::Saturating;
+use frame_support::traits::{fungibles::metadata::Inspect, ReservableCurrency};
+use parity_scale_codec::{Decode, Encode, MaxEncodedLen};
+use sp_core::U256;
+
+#[derive(Debug, Encode, Decode, scale_info::TypeInfo, PartialEq, MaxEncodedLen)]
+pub enum ForeignAssetMigrationStatus {
+	/// No migration in progress
+	Idle,
+	/// Migrating a foreign asset in progress
+	Migrating(ForeignAssetMigreationInfo),
+}
+
+impl Default for ForeignAssetMigrationStatus {
+	fn default() -> Self {
+		ForeignAssetMigrationStatus::Idle
+	}
+}
+
+#[derive(Debug, Encode, Decode, scale_info::TypeInfo, PartialEq, MaxEncodedLen)]
+pub(super) struct ForeignAssetMigreationInfo {
+	pub(super) asset_id: u128,
+	pub(super) remaining_balances: u32,
+	pub(super) remaining_approvals: u32,
+}
+
+impl<T: Config> Pallet<T>
+where
+	<T as pallet_assets::Config>::Balance: Into<U256>,
+	<T as pallet_asset_manager::Config>::ForeignAssetType: Into<Option<Location>>,
+{
+	/// Start a foreign asset migration by freezing the asset and creating the SC with the moonbeam
+	/// foreign assets pallet.
+	pub(super) fn do_start_foreign_asset_migration(
+		origin: OriginFor<T>,
+		asset_id: u128,
+	) -> DispatchResult {
+		ForeignAssetMigrationStatusValue::<T>::try_mutate(|status| -> DispatchResult {
+			ensure!(
+				*status == ForeignAssetMigrationStatus::Idle,
+				Error::<T>::MigrationNotFinished
+			);
+
+			// Freeze the asset
+			pallet_assets::Asset::<T>::try_mutate_exists(asset_id, |maybe_details| {
+				let details = maybe_details.as_mut().ok_or(Error::<T>::AssetNotFound)?;
+
+				details.status = pallet_assets::AssetStatus::Frozen;
+
+				let decimals = pallet_assets::Pallet::<T>::decimals(asset_id);
+				let symbol = pallet_assets::Pallet::<T>::symbol(asset_id)
+					.try_into()
+					.map_err(|_| Error::<T>::SymbolTooLong)?;
+				let name = <pallet_assets::Pallet<T> as Inspect<_>>::name(asset_id)
+					.try_into()
+					.map_err(|_| Error::<T>::NameTooLong)?;
+				let asset_type = pallet_asset_manager::AssetIdType::<T>::take(asset_id)
+					.ok_or(Error::<T>::AssetTypeNotFound)?;
+				let xcm_location: Location =
+					asset_type.into().ok_or(Error::<T>::LocationNotFound)?;
+
+				// Remove the precompile for the old foreign asset.
+				// Cleaning the precompile is done by removing the code and metadata
```

Comment:
This comment is wrong and should be rewrite, removing the AccountCodes and AccountCodesMetadata doesn't remove the precompile, and it's good this way, we should not remove the precompile before migration finalisation).
The precompile mapping is removed when we remove the asset data in pallet asset

Reply:
Indeed. I removed it.